functions/inputs/from.go

```diff
@@ -154,9 +155,141 @@ func (rule MergeFromRangeRule) Rewrite(node plan.PlanNode) (plan.PlanNode, bool)
 	merged, err := plan.MergePhysicalPlanNodes(node, from, fromRange)
 	if err != nil {
-		return node, false
+		return nil, false, err
 	}
-	return merged, true
+
+	return merged, true, nil
+}
+
+type MergeFromFilterRule struct {
+}
+
+func (MergeFromFilterRule) Name() string {
+	return "MergeFromFilterRule"
+}
+
+func (MergeFromFilterRule) Pattern() plan.Pattern {
+	return plan.Pat(transformations.FilterKind, plan.Pat(FromKind))
+}
+
+func (MergeFromFilterRule) Rewrite(filterNode plan.PlanNode) (plan.PlanNode, bool, error) {
+	filterSpec := filterNode.ProcedureSpec().(*transformations.FilterProcedureSpec)
+	bodyExpr, ok := filterSpec.Fn.Body.(semantic.Expression)
+	if !ok {
+		return filterNode, false, nil
+	}
+
+	pushable, notPushable, err := semantic.PartitionPredicates(bodyExpr, isPushableExpr)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if pushable == nil {
+		// Nothing could be pushed down, no rewrite can happen
+		return filterNode, false, nil
+	}
+
+	fromNode := filterNode.Predecessors()[0]
+	newFromSpec := fromNode.ProcedureSpec().Copy().(*FromProcedureSpec)
+	if newFromSpec.FilterSet {
+		newBody := semantic.ExprsToConjunction(newFromSpec.Filter.Body.(semantic.Expression), pushable)
+		newFromSpec.Filter.Body = newBody
+	} else {
+		newFromSpec.FilterSet = true
+		newFromSpec.Filter = filterSpec.Fn.Copy().(*semantic.FunctionExpression)
+		newFromSpec.Filter.Body = pushable
+	}
+
+	if notPushable == nil {
+		// All predicates could be pushed down, so eliminate the filter
+		mergedNode, err := plan.MergePhysicalPlanNodes(filterNode, fromNode, newFromSpec)
+		if err != nil {
+			return nil, false, err
+		}
+		return mergedNode, true, nil
+	}
+
+	err = fromNode.ReplaceSpec(newFromSpec)
+	if err != nil {
+		return nil, false, err
+	}
+
+	newFilterSpec := filterSpec.Copy().(*transformations.FilterProcedureSpec)
+	newFilterSpec.Fn.Body = notPushable
+	err = filterNode.ReplaceSpec(newFilterSpec)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return filterNode, true, nil
+}
+
+func isPushableExpr(expr semantic.Expression) (bool, error) {
+	switch e := expr.(type) {
+	case *semantic.LogicalExpression:
+		b, err := isPushableExpr(e.Left)
+		if err != nil {
+			return false, err
+		}
+
+		if !b {
+			return false, nil
+		}
+
+		return isPushableExpr(e.Right)
+
+	case *semantic.BinaryExpression:
+		if !isPushableOperator(e.Operator) {
+			return false, nil
+		}
+
+		b, err := isPushableExpr(e.Left)
+		if err != nil {
+			return false, err
+		}
+
+		if !b {
+			return false, nil
+		}
+
+		return isPushableExpr(e.Right)
+
+	case *semantic.StringLiteral:
+		return true, nil
+	case *semantic.IntegerLiteral:
+		return true, nil
+	case *semantic.BooleanLiteral:
+		return true, nil
+	case *semantic.FloatLiteral:
+		return true, nil
+	case *semantic.RegexpLiteral:
+		return true, nil
+	case *semantic.MemberExpression:
+		if idExpr, ok := e.Object.(*semantic.IdentifierExpression); ok && idExpr.Name == "r" {
```

Comment:
Could you modify this to capture the input argument to the filter function so that it works with functions that take an argument other that `r`. Such as `(x) => ...`.

Reply:
Sure I can do that.  I actually thought that it had to be named `r`.  But I just tried it with a different name and it worked.